### PR TITLE
add an Unapply typeclass

### PR DESCRIPTION
**Pull request by stew**
also adds traverseU / sequenceU and adds syntax for traverse.

### CODE REVIEW

**Comment by non** [comment]
Looks like a few style problems with whitespace.

File: core/src/main/scala/cats/Unapply.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+  // a function which will coerce the MA value into one of type M[A]
+  // this will end up being the identity function, but we can't supply
+  // it until we have proven that MA and M[A] are the same type
+  def subst: MA => M[A]
+}
+
+object Unapply {
+  // a convenience method for summoning Unapply instances
+  def apply[TC[_[_]], MA](implicit ev: Unapply[TC,MA]): Unapply[TC, MA] = implicitly
```
Comment (ceedubs) [inline core/src/main/scala/cats/Unapply.scala]: Should probably just return `ev` instead of calling `implicitly`.

**Comment by ceedubs** [comment]
One extremely minor comment, but :+1:. Thanks, @stew!

**Comment by non** [comment]
:+1:

**Pull request merged**